Mlem/App/Views/Shared/Toast/ToastType.swift

```diff
@@ -0,0 +1,115 @@
+//
+//  Toast.swift
+//  Mlem
+//
+//  Created by Sjmarf on 15/05/2024.
+//
+
+import MlemMiddleware
+import SwiftUI
+
+enum ToastType: Hashable {
+    case basic(
+        title: String,
+        subtitle: String? = nil,
+        systemImage: String? = nil,
+        color: Color,
+        duration: Double = 1.5
+    )
+    
+    case undoable(
+        title: String? = nil,
+        systemImage: String? = nil,
+        callback: () -> Void,
+        color: Color = .blue
+    )
+    
+    case error(_ details: ErrorDetails)
+    
+    case user(AnyUserProviding)
+    
+    var duration: Double {
+        switch self {
+        case let .basic(_, _, _, _, duration):
+            duration
+        case .undoable:
+            2.5
+        case .user:
+            1.0
+        case .error:
+            1.5
+        }
+    }
+    
+    var location: ToastLocation {
+        switch self {
+        case .undoable:
+            .bottom
+        default:
+            .top
+        }
+    }
+    
+    static func success(_ message: String? = nil) -> Self {
+        .basic(
+            title: message ?? "Success",
+            subtitle: nil,
+            systemImage: "checkmark.circle.fill",
+            color: .green,
+            duration: 1
+        )
+    }
+    
+    static func failure(_ message: String? = nil) -> Self {
+        .basic(
+            title: message ?? "Failed",
+            subtitle: nil,
+            systemImage: "xmark.circle.fill",
+            color: .red,
```

Comment:
(needs matching definition in `Palette`)

```suggestion
            color: Palette.main.failure,
```